app/policies/application_policy.rb

```diff
@@ -18,10 +18,15 @@ def resolve
     attr_reader :user, :scope
   end
 
-  attr_reader :user, :record
+  attr_reader :user, :api_key, :record
 
   def initialize(user, record)
-    @user = user
+    if user.is_a?(ApiKey)
```

Comment:
nit: make this a case statement, and raise if `user` isnt an ApiKey or User

Reply:
cc @Burgestrand, let's come up with a better way to make the policy polymorphic on the user type like we discussed at lunch today?

Reply:
I had started with [this approach](https://github.com/varvet/pundit?tab=readme-ov-file#additional-context) in the pundit docs, but it was more cumbersome than just noticing it's an api key. The alternative involved forwarding methods which didn't easily save me from checking if the method was available anyway.

Reply:
From what I understand the policies will end up being very different if it's an ApiKey versus User, to the point where it might be worth having entirely different policies for the two?

Pseudo-code, to show you what I'm thinking. Terse because I'm on my phone on a boat so it's hard to type 😅

```ruby
class ApplicationPolicy; end
class Users::PostPolicy < ApplicationPolicy; end
class ApiKeys::PostPolicy < ApplicationPolicy; end
```

Effectively separating the two entirely. You could use inheritance or composition to share behaviour between the boundary if you need to.

I believe this might even be doable using namespaces, so you might not need to adjust policy lookup behaviour which was my initial thought @segiddins

Reply:
Yeah, I considered namespaces. We use that for the admin policies which are all completely different than the userland policies. Usually the api key policies will just be an extra check, so inheritance could make sense: `api.can_do_that? && super`

Reply:
Gotcha! Yes, it can definitely make sense!

I don't have a solid enough grasp of what you're designing to have a strong opinion so all I'm providing are gut reactions based off of when @segiddins and I talked at Baltic Ruby 😊 

One personal worry is how easy it is for a programmer error to slip through. How easy is it to forget to add the extra check, and how big of an impact that has.

Something else that I figured was worth mentioning is that we've in the past composed policies together. It does make the policies more difficult to reason about so I don't strictly recommend this approach, but it's another tool in the bag.

```ruby
class ApiKeySpecific::SomeResourcePolicy
  def initialize(api_key, record) = # ...

  def update?
    return false unless api_key.can_configure_trusted_publishers?
    
    Pundit.policy(api_key&.user, record).update?
  end
end
```